rust/worker/src/execution/operators/brute_force_knn.rs

```diff
@@ -17,45 +21,103 @@ pub struct BruteForceKnnOperator {}
 
+#[derive(Debug)]
+struct Entry {
+    index: usize,
+    distance: f32,
+}
+
+impl Ord for Entry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.distance == other.distance {
+            Ordering::Equal
+        } else if self.distance > other.distance {
+            // This is a min heap, so we need to reverse the ordering.
+            Ordering::Less
+        } else {
+            // This is a min heap, so we need to reverse the ordering.
+            Ordering::Greater
+        }
+    }
+}
+
+impl PartialOrd for Entry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Entry {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+
+impl Eq for Entry {}
+
 #[async_trait]
 impl Operator<BruteForceKnnOperatorInput, BruteForceKnnOperatorOutput> for BruteForceKnnOperator {
     type Error = ();
 
     async fn run(&self, input: &BruteForceKnnOperatorInput) -> BruteForceKnnOperatorResult {
-        // We could use a heap approach here, but for now we just sort the distances and take the
-        // first k.
-        let mut sorted_indices_distances = input
-            .data
-            .iter()
-            .map(|data| input.distance_metric.distance(&input.query, data))
-            .enumerate()
-            .collect::<Vec<(usize, f32)>>();
-        sorted_indices_distances.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
-        let (sorted_indices, sorted_distances) = sorted_indices_distances
-            .drain(..cmp::min(input.k, input.data.len()))
-            .unzip();
+        let mut heap = BinaryHeap::with_capacity(input.k);
+        let data_chunk = &input.data;
+        for data in data_chunk.iter() {
+            let embedding_record = data.0;
+            let index = data.1;
 
+            let embedding = match &embedding_record.embedding {
+                Some(embedding) => embedding,
+                None => {
+                    continue;
+                }
+            };
+            let distance = input.distance_metric.distance(&embedding[..], &input.query);
+            heap.push(Entry { index, distance });
+        }
+
+        let mut visibility = vec![false; data_chunk.len()];
+        let mut sorted_indices = Vec::with_capacity(input.k);
+        let mut sorted_distances = Vec::with_capacity(input.k);
+        let mut i = 0;
+        while i < input.k {
+            let entry = match heap.pop() {
+                Some(entry) => entry,
+                None => {
+                    break;
+                }
+            };
+            sorted_indices.push(entry.index);
+            sorted_distances.push(entry.distance);
+            visibility[entry.index] = true;
+            i += 1;
+        }
+        let mut data_chunk = data_chunk.clone();
+        data_chunk.set_visibility(visibility);
         Ok(BruteForceKnnOperatorOutput {
+            data: data_chunk,
```

Comment:
Just for my understanding - we mask out everything except that in the result set and we expect the user to leverage the sorted_indices in order to index into data_chunk and gather the resutls yeah?